File: AndroidStudioProjects/DaggerDemoJava/app/src/main/java/com/example/myapplication/WheelsModule.java
package com.example.myapplication;

import android.util.Log;

import dagger.Module;
import dagger.Provides;

@Module
public class WheelsModule {

    private static final String TAG = "WheelsModule";
    @Provides
    Rims providesRims(){
        Log.d(TAG, "providesRims: ");
        return new Rims();
    }
    @Provides
    Tires providesTires(){
        Log.d(TAG, "providesTires: ");
        return new Tires();
    }

    @Provides
    Wheels providesWheels(Rims rims, Tires tires){
        Log.d(TAG, "providesWheels: ");
        return new Wheels(rims,tires);
    }
}

File: AndroidStudioProjects/DaggerDemoKotlin/app/src/main/java/com/example/daggerdemokotlin/WheelsModule.kt
package com.example.daggerdemokotlin

import dagger.Module
import dagger.Provides

@Module
class WheelsModule {

    @Provides
    fun providesRims(): Rims{
        System.out.println("rims created")
        return Rims()
    }

    @Provides
    fun providesTires():Tires{
        System.out.println("tires created")
        return Tires()
    }

    @Provides
    fun providesWheels(rims: Rims,tires: Tires):Wheels{
        System.out.println("Wheels module")
        return Wheels(rims,tires)
    }
}